libvast/test/fixtures/events.cpp

```diff
@@ -92,6 +96,77 @@ events::copy(std::vector<table_slice_handle> xs) {
   return result;
 }
 
+class event_tracking_builder {
+  table_slice_builder_ptr inner_;
+  std::vector<event*> memory_;
+
+public:
+  event_tracking_builder() {
+  }
+
+  explicit event_tracking_builder(table_slice_builder_ptr b) : inner_{b} {
+  }
+
+  bool add(event* e) {
+    if (!inner_->add(e->timestamp()))
+      FAIL("builder->add() failed");
+    if (!inner_->recursive_add(e->data(), e->type()))
+      FAIL("builder->recursive_add() failed");
+
+    memory_.push_back(e);
+    return true;
+  }
+
+  auto rows() const {
+    return inner_->rows();
+  }
+
+  const std::vector<event*> entries() const {
+    return memory_;
+  }
+
+  table_slice_handle finish() {
+    memory_.clear();
+    return inner_->finish();
+  }
+};
+
+
+/// Tries to access the builder for `layout`.
+class builders {
+  using Map = std::map<std::string, event_tracking_builder>;
+
+  Map builders_;
+
+public:
+  event_tracking_builder* get(const type& layout) {
+    auto i = builders_.find(layout.name());
+    if (i != builders_.end())
+      return &i->second;
+    return caf::visit(
+      detail::overload(
+        [&](const record_type& rt) -> event_tracking_builder* {
+          // We always add a timestamp as first column to the layout.
+          auto internal = rt;
+          record_field tstamp_field{"timestamp", timestamp_type{}};
+          internal.fields.insert(internal.fields.begin(),
+                                 std::move(tstamp_field));
+          auto& ref = builders_[layout.name()];
```

Comment:
This could be more straightforward when using `emplace` instead:

```c++
event_tracking_builder tmp{default_table_slice::make_builder(std::move(internal))};
return &(*builder_.emplace(layout.name(), std::move(tmp)).first);
```